File: driver/dac.c
/* j4cDAC driver
 *
 * Copyright 2009 Andrew Kibler
 * Copyright 2011 Jacob Potter
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of either the GNU General Public License version 2
 * or 3, or the GNU Lesser General Public License version 3, as published
 * by the Free Software Foundation, at your option.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#define WINVER 0x0501

#include <stdio.h>
#include <math.h>
#include <string.h>

#include "dac.h"

#define DEBUG_THRESHOLD		800

extern dac_t * dac_list;

unsigned __stdcall LoopUpdate(void *d);

/* Initialize a dac_t */
int dac_init(dac_t *d) {
	memset(d, 0, sizeof(*d));
	InitializeCriticalSection(&d->buffer_lock);
	return 0;
}

/* Connect */
int dac_open_connection(dac_t *d) {
	char host[40];
	strncpy(host, inet_ntoa(d->addr), sizeof(host) - 1);
	host[sizeof(host) - 1] = 0;

	// Initialize buffer
	d->buffer_read = 0;
	d->buffer_fullness = 0;
	memset(d->buffer, sizeof(d->buffer), 0);

	// Connect to the DAC
	if (dac_connect(d, host, "7765") < 0) {
		flogd(d, "!! DAC connection failed.\n");
		return -1;
	}

	// Fire off the worker thread
	d->loop_go = CreateEvent(NULL, FALSE, FALSE, NULL);
	d->state = ST_READY;

	d->workerthread = (HANDLE)_beginthreadex(NULL, 0, &LoopUpdate, d, 0, NULL);
	if (!d->workerthread) {
		flogd(d, "!! Begin thread error: %s\n", strerror(errno));
		return -1;
	}

	flogd(d, "Ready.\n");

	return 0;
}

void dac_close_connection(dac_t *d) {
	EnterCriticalSection(&d->buffer_lock);
	if (d->state == ST_READY)
		SetEvent(d->loop_go);
	d->state = ST_SHUTDOWN;
	LeaveCriticalSection(&d->buffer_lock);

	int rv = WaitForSingleObject(d->workerthread, 250);
	if (rv != WAIT_OBJECT_0)
		flogd(d, "Exit thread timed out.\n");

	if (d->state == ST_READY) {
		CloseHandle(d->workerthread);
	}

	dac_disconnect(d);
	d->state = ST_DISCONNECTED;
}

/* Look up a DAC by index or unique-ID
 */
dac_t * dac_get(int num) {
	dac_t *d = dac_list;
	if (num >= 0) {
		while (num--) {
			if (!d->next) {
				/* If they pass one past the last DAC,
				 * just return the last one. Ugh, buggy
				 * software. */
				if (!num) return d;
				else return NULL;
			}
			d = d->next;
		}
	} else {
		while (d) {
			if (num == d->dac_id) break;
			d = d->next;
		}
	}

	return d;
}

void dac_get_name(dac_t *d, char *buf, int max) {
	snprintf(buf, max, "Ether Dream %02x%02x%02x",
		d->mac_address[3], d->mac_address[4], d->mac_address[5]);
}

/* Buffer access
 */
struct buffer_item *buf_get_write(dac_t *d) {
	EnterCriticalSection(&d->buffer_lock);
	int write = (d->buffer_read + d->buffer_fullness) % BUFFER_NFRAMES;
	LeaveCriticalSection(&d->buffer_lock);

	flogd(d, "M: Writing to index %d\n", write);

	return &d->buffer[write];
}

void buf_advance_write(dac_t *d) {
	EnterCriticalSection(&d->buffer_lock);
	d->buffer_fullness++;
	if (d->state == ST_READY)
		SetEvent(d->loop_go);
	d->state = ST_RUNNING;
	LeaveCriticalSection(&d->buffer_lock);
}

int dac_get_status(dac_t *d) {
	EnterCriticalSection(&d->buffer_lock);
	int fullness = d->buffer_fullness;
	LeaveCriticalSection(&d->buffer_lock);

	if (fullness == BUFFER_NFRAMES) {
		return GET_STATUS_BUSY;
	} else {
		return GET_STATUS_READY;
	}
}

/* Write a frame.
 */

int do_write_frame(dac_t *d, const void * data, int bytes, int pps,
	int reps, int (*convert)(struct buffer_item *, const void *, int)) {

	int points = convert(NULL, NULL, bytes);

	if (reps == ((uint16_t) -1))
		reps = -1;

	/* If not ready for a new frame, bail */
	if (dac_get_status(d) != GET_STATUS_READY) {
		flogd(d, "M: NOT READY: %d points, %d reps\n", points, reps);
		return 0;
	}

	/* Ignore 0-repeat frames */
	if (!reps)
		return 1;

	int internal_reps = 250 / points;
	char * bigdata = NULL;
	if (internal_reps) {
		bigdata = malloc(bytes * internal_reps);
		int i;
		for (i = 0; i < internal_reps; i++) {
			memcpy(bigdata + i*bytes, data, bytes);
		}
		bytes *= internal_reps;
		data = bigdata;
	}

	flogd(d, "M: Writing: %d/%d points, %d reps, %d pps\n",
		points, convert(NULL, NULL, bytes), reps, pps);

	struct buffer_item *next = buf_get_write(d);
	convert(next, data, bytes);
	next->pps = pps;
	next->repeatcount = reps;

	buf_advance_write(d);

	if (bigdata) free(bigdata);

	return 1;
}

#include <locale.h>

// This is the buffer filling thread for WriteFrame() and WriteFrameNR()
unsigned __stdcall LoopUpdate(void *dv){
	dac_t *d = (dac_t *)dv;

	int res;
	res = SetThreadPriority(GetCurrentThread, THREAD_PRIORITY_TIME_CRITICAL);
	flogd(d, "SetThreadPriority ret %d\n");

	if (timeBeginPeriod(1) == TIMERR_NOERROR) {
		flogd(d, "Timer set to 1ms\n");
	} else {
		flogd(d, "Timer error\n");
	}

	EnterCriticalSection(&d->buffer_lock);

	while (1) {
		/* Wait for us to have data */
		while (d->state == ST_READY) {
			LeaveCriticalSection(&d->buffer_lock);

			flogd(d, "L: Waiting...\n");

			int res = WaitForSingleObject(d->loop_go, INFINITE);

			EnterCriticalSection(&d->buffer_lock);
			if (res != WAIT_OBJECT_0) {
				flogd(d, "!! WFSO failed: %lu\n",
					GetLastError());
				d->state = ST_BROKEN;
				timeEndPeriod(1);
				return 0;
			}
		}

		if (d->state != ST_RUNNING) {
			flogd(d, "L: Shutting down.\n");
			LeaveCriticalSection(&d->buffer_lock);
			timeEndPeriod(1);
			return 0;
		}

		LeaveCriticalSection(&d->buffer_lock);

		struct buffer_item *b = &d->buffer[d->buffer_read];
		int cap;
		int iters = 0;

		while (1) {
			/* Estimate how much data has been consumed since the last
			 * time we got an ACK. */
			LARGE_INTEGER time;
			QueryPerformanceCounter(&time);
			LONGLONG time_diff = time.QuadPart - d->conn.last_ack_time.QuadPart;
			LONGLONG points_used = (time_diff * b->pps) / timer_freq.QuadPart;
			int iu = points_used;
			if (d->conn.resp.dac_status.playback_state != 2) iu = 0;

			int expected_fullness = d->conn.resp.dac_status.buffer_fullness \
				- iu + d->conn.unacked_points;

			/* Now, see how much data we should write. */
			cap = 1700 - expected_fullness;

			if (d->conn.resp.dac_status.buffer_fullness < DEBUG_THRESHOLD
			    || expected_fullness < DEBUG_THRESHOLD)
				flogd(d, "L: b %d + %d - %d = %d, w %d om %d st %d\n",
					d->conn.resp.dac_status.buffer_fullness,
					d->conn.unacked_points,
					iu, expected_fullness, cap,
					d->conn.pending_meta_acks,
					d->conn.resp.dac_status.playback_state);

			if (cap > DAC_MIN_SEND) break;
			if (iters++ > 1) break;

			/* Wait a little. */
			int diff = DAC_MIN_SEND - cap;
			int sleeptime = 1 + (1000 * diff / b->pps);
			Sleep(sleeptime);
		}

		/* How many points can we send? */
		int b_left = b->points - b->idx;

		if (cap > b_left)
			cap = b_left;
		if (cap > 80)
			cap = 80;
		if (cap < 0)
			cap = 1;

		int res = dac_send_data(d, b->data + b->idx, cap, b->pps);

		if (res < 0) {
			/* Welp, something's wrong. There's not much we
			 * can do at an API level other than start throwing
			 * "error" returns up to the application... */
			EnterCriticalSection(&d->buffer_lock);
			d->state = ST_BROKEN;
			LeaveCriticalSection(&d->buffer_lock);
			timeEndPeriod(1);
			return 1;
		}

		/* What next? */
		EnterCriticalSection(&d->buffer_lock);
		b->idx += res;

		if (b->idx < b->points) {
			/* There's more in this frame. */
			continue;
		}

		b->idx = 0;

		if (b->repeatcount > 1) {
			/* Play this frame again? */
			b->repeatcount--;
		} else if (d->buffer_fullness > 1) {
			/* Move to the next frame */
			flogd(d, "L: advancing frame - buffer fullness %d\n",
				d->buffer_fullness);
			d->buffer_fullness--;
			d->buffer_read++;
			if (d->buffer_read >= BUFFER_NFRAMES)
				d->buffer_read = 0;
		} else if (b->repeatcount >= 0) {
			/* Stop playing until we get a new frame. */
			flogd(d, "L: returning to idle\n");
			d->state = ST_READY;
		} else {
			/* Repeat this frame */
		}

		/* If we get here without hitting any of the above cases,
		 * then repeatcount is negative and there's no new frame -
		 * so we're just supposed to keep playing this one again
		 * and again. */ 
	}

	timeEndPeriod(1);
	return 0;
}

File: tools/tester/dac.py
# j4cDAC test code
#
# Copyright 2011 Jacob Potter
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import socket
import time
import struct

def pack_point(x, y, r, g, b, i = -1, u1 = 0, u2 = 0, flags = 0):
	"""Pack some color values into a struct dac_point.

	Values must be specified for x, y, r, g, and b. If a value is not
	passed in for the other fields, i will default to max(r, g, b); the 
	rest default to zero.
	"""
	
	if i < 0:
		i = max(r, g, b)

	return struct.pack("<HhhHHHHHH", flags, x, y, r, g, b, i, u1, u2)


class ProtocolError(Exception):
	"""Exception used when a protocol error is detected."""
	pass


class Status(object):
	"""Represents a status response from the DAC."""

	def __init__(self, data):
		"""Initialize from a chunk of data."""
		self.protocol_version, self.le_state, self.playback_state, \
		  self.source, self.le_flags, self.playback_flags, \
		  self.source_flags, self.fullness, self.point_rate, \
		  self.point_count = \
			struct.unpack("<BBBBHHHHII", data)

	def dump(self, prefix = " - "):
		"""Dump to a string."""
		lines = [
			"Light engine: state %d, flags 0x%x" %
				(self.le_state, self.le_flags),
			"Playback: state %d, flags 0x%x" %
				(self.playback_state, self.playback_flags),
			"Buffer: %d points" %
				(self.fullness, ),
			"Playback: %d kpps, %d points played" %
				(self.point_rate, self.point_count),
			"Source: %d, flags 0x%x" %
				(self.source, self.source_flags)
		]
		for l in lines:
			print prefix + l


class BroadcastPacket(object):
	"""Represents a broadcast packet from the DAC."""

	def __init__(self, st):
		"""Initialize from a chunk of data."""
		self.mac = st[:6]
		self.hw_rev, self.sw_rev, self.buffer_capacity, \
		self.max_point_rate = struct.unpack("<HHHI", st[6:16])
		self.status = Status(st[16:36])

	def dump(self, prefix = " - "):
		"""Dump to a string."""
		lines = [
			"MAC: " + ":".join(
				"%02x" % (ord(o), ) for o in self.mac),
			"HW %d, SW %d" %
				(self.hw_rev, self.sw_rev),
			"Capabilities: max %d points, %d kpps" %
				(self.buffer_capacity, self.max_point_rate)
		]
		for l in lines:
			print prefix + l
		self.status.dump(prefix)


class DAC(object):
	"""A connection to a DAC."""

	def read(self, l):
		"""Read exactly length bytes from the connection."""
		while l > len(self.buf):
			self.buf += self.conn.recv(4096)

		obuf = self.buf
		self.buf = obuf[l:]
		return obuf[:l]

	def readresp(self, cmd):
		"""Read a response from the DAC."""
		data = self.read(22)
		response = data[0]
		cmdR = data[1]
		status = Status(data[2:])

#		status.dump()

		if cmdR != cmd:
			raise ProtocolError("expected resp for %r, got %r"
				% (cmd, cmdR))

		if response != "a":
			raise ProtocolError("expected ACK, got %r"
				% (response, ))

		self.last_status = status
		return status

	def __init__(self, host, port = 7765):
		"""Connect to the DAC over TCP."""
		conn = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		conn.connect((host, port))
		self.conn = conn
		self.buf = ""

		# Read the "hello" message
		first_status = self.readresp("?")
		first_status.dump()

	def begin(self, lwm, rate):
		cmd = struct.pack("<cHI", "b", lwm, rate)
		self.conn.sendall(cmd)
		return self.readresp("b")

	def update(self, lwm, rate):
		cmd = struct.pack("<cHI", "u", lwm, rate)
		self.conn.sendall(cmd)
		return self.readresp("u")

	def encode_point(self, point):
		return pack_point(*point)

	def write(self, points):
		epoints = map(self.encode_point, points)
		cmd = struct.pack("<cH", "d", len(epoints))
		self.conn.sendall(cmd + "".join(epoints))
		return self.readresp("d")

	def prepare(self):
		self.conn.sendall("p")
		return self.readresp("p")

	def stop(self):
		self.conn.sendall("s")
		return self.readresp("s")

	def estop(self):
		self.conn.sendall("\xFF")
		return self.readresp("\xFF")

	def clear_estop(self):
		self.conn.sendall("c")
		return self.readresp("c")

	def ping(self):
		self.conn.sendall("?")
		return self.readresp("?")

	def play_stream(self, stream):
		# First, prepare the stream
		if self.last_status.playback_state == 2:
			raise Exception("already playing?!")
		elif self.last_status.playback_state == 0:
			self.prepare()

		started = 0

		while True:
			# How much room?
			cap = 1799 - self.last_status.fullness
			points = stream.read(cap)

			if cap < 100:
				time.sleep(0.005)
				cap += 150

#			print "Writing %d points" % (cap, )
			t0 = time.time()
			self.write(points)
			t1 = time.time()
#			print "Took %f" % (t1 - t0, )

			if not started:
				self.begin(0, 30000)
				started = 1


def find_dac():
	"""Listen for broadcast packets."""

	s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
	s.bind(("0.0.0.0", 7654))

	while True:
		data, addr = s.recvfrom(1024)
		bp = BroadcastPacket(data)
		
		print "Packet from %s: " % (addr, )
		bp.dump()

def find_first_dac():
	s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
	s.bind(("0.0.0.0", 7654))
	data, addr = s.recvfrom(1024)
	bp = BroadcastPacket(data)
	print "Packet from %s: " % (addr, )
	return addr[0]
